keras/src/backend/jax/distribution_lib.py

```diff
@@ -118,72 +118,64 @@ def distribute_data_input(inputs, layout):
             `jax.sharding.Sharding` instance.
 
     Returns:
-        Distributed inputs thats been properly put to local devices.
+        A global batch distributed according to `layout`.
     """
     if not isinstance(layout, jax.sharding.Sharding):
         layout = _to_jax_layout(layout)
-    if layout.is_fully_addressable:
-        return jax.device_put(inputs, layout)
-
-    # We need the jax mesh information to determine how to place the data
-    # on to each of the worker.
-    jax_mesh = layout.mesh
-    mesh_rank = len(jax_mesh.shape)
-    per_process_batch_size = inputs.shape[0]
-    if mesh_rank == 1:
-        # This is data parallel mesh only. We will split the full data
-        # across the batch dim.
-        num_split = jax.local_device_count()
-        per_replica_batch_size = per_process_batch_size // num_split
+
+    mesh_shape = list(layout.mesh.shape.values())
+    mesh_batch_dim_size = mesh_shape[0]
+    num_model_replicas_total = mesh_batch_dim_size
+    mesh_model_dim_size = mesh_shape[1] if len(mesh_shape) > 1 else 1
+    num_model_replicas_per_process = num_model_replicas_total / num_processes()
```

Comment:
Should this be `//` integer division? What happens if there are floating point values. Should that raise ValueError or is it OK?

Reply:
We can have a fraction of a model replica on a process, that's why it's `/` instead of `//`.
For example, if we want to shard the model into 4 pieces but each of our processes have only 2 devices, we have to spread our model across 2 processes.

Reply:
Discussed offline. We may need to check that its integer divisible, otherwise `np.split` that uses this value would give an error. Samaneh plans to create a PR with a fix and check for integer divisibility.